includes/Core/Authentication/Clients/OAuth_Client.php

```diff
@@ -205,16 +205,22 @@ public function get_client() {
 		}
 
 		$token = array(
-			'access_token' => $access_token,
-			'expires_in'   => $this->user_options->get( self::OPTION_ACCESS_TOKEN_EXPIRES_IN ),
-			'created'      => $this->user_options->get( self::OPTION_ACCESS_TOKEN_CREATED ),
+			'access_token'  => $access_token,
+			'expires_in'    => $this->user_options->get( self::OPTION_ACCESS_TOKEN_EXPIRES_IN ),
+			'created'       => $this->user_options->get( self::OPTION_ACCESS_TOKEN_CREATED ),
+			'refresh_token' => $this->get_refresh_token(),
 		);
-		if ( ! $this->using_proxy() ) {
-			$token['refresh_token'] = $this->get_refresh_token();
-		}
 
 		$this->google_client->setAccessToken( $token );
 
+		// This is called when the client refreshes the access token on-the-fly.
+		$this->google_client->setTokenCallback(
+			function( $cache_key, $access_token ) {
+				// All we can do here is assume an hour as it usually is.
+				$this->set_access_token( $access_token, HOUR_IN_SECONDS );
+			}
+		);
+
 		// If the token expired or is going to expire in the next 30 seconds.
 		if ( $this->google_client->isAccessTokenExpired() ) {
 			$this->refresh_token();
```

Comment:
Is this needed now that the `refresh_token` is always set? Doesn't it refresh automatically as needed?

Reply:
It should, but I think the manual method is still more secure for now and doesn't interfere with an actual request being made (which would be especially problematic under a very possible race condition where we fire multiple API requests to our own REST API at the same time, which would trigger their own Google API requests).

Again, the whole change to `Google_Proxy_Client` is just for some edge case and shouldn't adjust how we manually refresh the token. We can definitely revisit this in the future based on some thorough testing.